### ISSUE TITLE
Accessing hidden prompt buffer stops insert mode unexpectedly

### ISSUE THREAD

**Issue by jedrzejboczar** [neovim/neovim#22924]
### Problem

First of all, I wanted to debug more before creating the issue, but noticed the 0.9 release so I'm creating it now.

Whenever I start nvim-dap debug session Neovim stops entering insert mode. Whenever I try to enter insert mode, e.g. by pressing `a`, it immediately exists. If I try to hold the `a` key once in a while a character will be inserted, but generally the insert-exit happens pretty fast.

I tried to pinpoint the issue by checking different plugins, but couldn't find a related change. Then I did `git bisect` on Neovim repo and found that the **commit that introduced the issue** was https://github.com/neovim/neovim/commit/10baf89712724b4b95f7c641f2012f051737003c. Judging from the description it is related to insert mode so it makes sense.  

I didn't yet have time to debug it more and may not have enough time soon so wanted to report it here.

### Steps to reproduce

Will try to provide detailed repro when I have more time to debug this. For reference, I'm starting nvim-dap debug session with cppdebug, I have disabled nvim-dap-ui. I am using noice.nvim if that may be related (juding from the vim patch that introduced the error it might be related to some noice prompt buffers?).

### Expected behavior

Can enter insert mode.

### Neovim version (nvim -v)

nightly 10baf89712724b4b95f7c641f2012f051737003c

### Vim (not Nvim) behaves the same?

not applicable

### Operating system/version

Arch Linux

### Terminal name/version

alacritty 0.12.0 (5a728195)

### $TERM environment variable

xterm-256color

### Installation

build from repo

**Comment by jedrzejboczar** [comment]
After some debugging I found that this happens only when there is a hidden, modified prompt buffer `buftype=prompt`. But with `nvim --clean` the bug doesn't happen. I narrowed down the bug to usage of statusline component from [feline.nvim](https://github.com/freddiehaddad/feline.nvim) - enabling `file_info` component with option `type='unique-short'`. 

I created a minimal reproduction config with no dependencies, with the feline component code extracted. Run it as 
```sh
{ nvim --clean -u repro.lua repro.lua && echo 'ok'; } || echo 'FAIL'
```
or remove the automatic exit to check it yourself. Removing the line `filename = get_unique_filename(filename, true)` fixes the problem, but I'm not sure what exactly causes the behaviour.

File `repro.lua`:
```lua
---------------------------------------------------------------------------------
-- Functions extracted from feline.nvim for provider:
--   provider = { name = 'file_info', opts = { type = 'unique-short' } }

-- Get the names of all current listed buffers
local function get_current_filenames()
    local listed_buffers = vim.tbl_filter(function(bufnr)
        return vim.bo[bufnr].buflisted and vim.api.nvim_buf_is_loaded(bufnr)
    end, vim.api.nvim_list_bufs())

    return vim.tbl_map(vim.api.nvim_buf_get_name, listed_buffers)
end

-- Get unique name for the current buffer
local function get_unique_filename(filename, shorten)
    local filenames = vim.tbl_filter(function(filename_other)
        return filename_other ~= filename
    end, get_current_filenames())

    if shorten then
        filename = vim.fn.pathshorten(filename)
        filenames = vim.tbl_map(vim.fn.pathshorten, filenames)
    end

    -- Reverse filenames in order to compare their names
    filename = string.reverse(filename)
    filenames = vim.tbl_map(string.reverse, filenames)

    local index

    -- For every other filename, compare it with the name of the current file char-by-char to
    -- find the minimum index `i` where the i-th character is different for the two filenames
    -- After doing it for every filename, get the maximum value of `i`
    if next(filenames) then
        index = math.max(unpack(vim.tbl_map(function(filename_other)
            for i = 1, #filename do
                -- Compare i-th character of both names until they aren't equal
                if filename:sub(i, i) ~= filename_other:sub(i, i) then
                    return i
                end
            end
            return 1
        end, filenames)))
    else
        index = 1
    end

    -- Iterate backwards (since filename is reversed) until a "/" is found
    -- in order to show a valid file path
    while index <= #filename do
        if filename:sub(index, index) == '/' then
            index = index - 1
            break
        end

        index = index + 1
    end

    return string.reverse(string.sub(filename, 1, index))
end

---------------------------------------------------------------------------------
-- Configuration

function _G.Statusline()
  local filename = vim.api.nvim_buf_get_name(0)
  filename = get_unique_filename(filename, true)
  return 'File: ' .. filename
end

vim.opt.statusline = "%{%v:lua.Statusline()%}"

---------------------------------------------------------------------------------
-- Automated test
local function co_resume()
    local co = assert(coroutine.running())
    return function(...)
        coroutine.resume(co, ...)
    end
end

local function sleep(ms)
    vim.defer_fn(co_resume(), ms)
    coroutine.yield()
end

local function feed_keys(keys)
    vim.api.nvim_feedkeys(
        vim.api.nvim_replace_termcodes(keys, true, false, true),
        'n',
        false
    )
end

local function stop()
    sleep(100)
end

coroutine.wrap(function()
    -- Create new buffer
    stop()
    vim.cmd.new()
    vim.bo.buftype = 'prompt'
    vim.bo.bufhidden = 'hide'

    -- Enter some text in prompt buffer
    stop()
    feed_keys('aSome text<cr><esc>')

    -- Hide the buffer
    stop()
    vim.cmd.hide()

    -- Try to enter insert mode
    stop()
    feed_keys('a')

    -- Check if it worked
    stop()
    local mode = vim.fn.mode()

    -- Make sure to exit insert mode if it worked
    stop()
    feed_keys('<esc>')

    -- -- Show result
    -- vim.cmd.enew()
    -- vim.api.nvim_buf_set_lines(0, -1, -1, true, {'After entering insert mode the mode was: ' .. mode})
    -- sleep(3000)

    -- Close all buffers
    for _, buf in ipairs(vim.api.nvim_list_bufs()) do
        vim.api.nvim_buf_delete(buf, { force = true })
    end

    -- Exit ok/error
    if mode == 'i' then
        vim.cmd.quit()
    else
        vim.cmd.cquit()
    end
end)()
```

**Comment by Shougo** [comment]
It has the side effect.

https://github.com/nvim-telescope/telescope.nvim/issues/2501
https://github.com/nvim-telescope/telescope.nvim/issues/2027#issuecomment-1538718111

**Comment by zeertzjq** [comment]
I think it was fixed in #23785.
